### PR TITLE
storage: fix race between context cancel & snapshots

### PR DESCRIPTION
**Pull request by cuongdo**
In (*RaftTransport).ProcessQueue(), the context for the gRPC calls is
being cancelled as part of a defer block. Unfortunately, errors (which
happen too frequently) cause the method to return and cancel outstanding
Raft requests, including snapshots, prematurely.

This change introduces a WaitGroup that blocks the cancellation until
the outstanding Raft requests have finished. This speeds up
up-replication by ~40-50%.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7252)

### CODE REVIEW

File: storage/raft_transport.go [hunk 1]
```diff
@@ -192,8 +192,18 @@ func (t *RaftTransport) processQueue(nodeID roachpb.NodeID) {
...
-	defer cancel()
+	defer func() {
+		// After pending Raft requests have finished, cancel the Context to free
+		// gRPC resources.
+		wg.Wait()
+		if log.V(1) {
+			log.Infof("cancelling context")
+		}
+		cancel()
+	}()
```
Comment (petermattis) [inline storage/raft_transport.go]: This seems odd. From looking at the code I would think we are trying to explicitly cancel inflight raft requests when we return. What code path is being taken that we're returning from this method but don't want to cancel inflight raft requests?

File: storage/raft_transport.go [hunk 2]
```diff
@@ -255,9 +265,11 @@ func (t *RaftTransport) processQueue(nodeID roachpb.NodeID) {
 			}
 			return
 		case req := <-ch:
+			wg.Add(1)
```
Comment (petermattis) [inline storage/raft_transport.go]: I think you only need to do the `wg.{Add,Done}` dance when `req.Message.Type == raftpb.MsgSnap`. In the other branch it is impossible for the `defer` to be running concurrently with `restStream.Send`.

File: storage/raft_transport.go (hunk 1 above)
Comment (cuongdo) [inline storage/raft_transport.go]: There are gRPC protobuf unmarshal errors that are causing this function to return when there are one or more in-flight snapshot requests. There's also an as-of-yet unexplained context cancellation that's affecting some snapshot requests, which also causes the function to return.

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/raft_transport.go, line 203 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKLJTcEewdJF79ppMlI:-KKLJTcEewdJF79ppMlJ:771841171) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L203)):*
> ```Go
> 		wg.Wait()
> 		if log.V(1) {
> 			log.Infof("cancelling context")
> ```

log more detail here, perhaps `log.Infof("closing Raft transport stream to node %d at %s" nodeID, addr)`

---

*[storage/raft_transport.go, line 206 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKKmkuz4o1XEuYpKS2u-r1-206:-KKLJsJfhrSBjt-8Woae:1755484025) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L206)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> There are gRPC protobuf unmarshal errors that are causing this function to return when there are one or more in-flight snapshot requests. There's also an as-of-yet unexplained context cancellation that's affecting some snapshot requests, which also causes the function to return.
</details>
@petermattis the intention is not to cancel inflight requests - it is to close the stream when we return.

---

*[storage/raft_transport.go, line 268 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKKmkuz4o1XEuYpKS2u-r1-268:-KKLJgqcF6Hux8wKjRv9:-1530771663) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L268)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think you only need to do the `wg.{Add,Done}` dance when `req.Message.Type == raftpb.MsgSnap`. In the other branch it is impossible for the `defer` to be running concurrently with `restStream.Send`.
</details>
Agreed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252)*

**Comment by cuongdo** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/raft_transport.go, line 203 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKLJTcEewdJF79ppMlI:-KKLZOO_LKbafl-WR-Kh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L203)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> log more detail here, perhaps `log.Infof("closing Raft transport stream to node %d at %s" nodeID, addr)`
</details>
Done.

---

*[storage/raft_transport.go, line 206 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKKmkuz4o1XEuYpKS2u-r1-206:-KKLZP6Hq4tTrzUMOqen:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L206)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> @petermattis the intention is not to cancel inflight requests - it is to close the stream when we return.
</details>
Done.

---

*[storage/raft_transport.go, line 268 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKKmkuz4o1XEuYpKS2u-r1-268:-KKLZPipOS-3ycCZj7LE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L268)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Agreed.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/raft_transport.go, line 206 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKKmkuz4o1XEuYpKS2u-r1-206:-KKLaZmo4jX7seKoSBKX:863495976) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L206)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Done.
</details>
Got it, but it still isn't clear to me why this change is helping. @cdo Do you mind tracking down what is causing the gRPC protobuf unmarshal errors first? I worry that this is covering up the real error and not fixing it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252)*

**Comment by cuongdo** [comment]
I don't think this change is covering up anything; it simply reduces the
impact of the unmarshal error. That said, I was planning to look at the
unmarshal errors next anyway, They're happening reliably every couple of
seconds during the 1->3 node allocator test.

**Comment by petermattis** [comment]
I guess covering up might be the wrong phrasing. If the unmarshal error was not happening this change wouldn't be necessary. On the other hand, I don't see a particular problem with merging this if the unmarshal error is still easily reproducible.

:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-:-KKLeN9ZpD6vwZvG9dCG:2146989179)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/raft_transport.go, line 203 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKLJTcEewdJF79ppMlI:-KKLeuchv9xL0gJVEHf5:-980550941) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L203)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Done.
</details>
nit: "due to return" isn't really helpful, is it? there are no other reasons that the stream is closed.

---

*[storage/raft_transport.go, line 206 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKKmkuz4o1XEuYpKS2u-r1-206:-KKLf2WYwtA2lNH4iB0a:1966880128) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L206)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Got it, but it still isn't clear to me why this change is helping. @cdo Do you mind tracking down what is causing the gRPC protobuf unmarshal errors first? I worry that this is covering up the real error and not fixing it. 
</details>
This isn't really covering it up - those errors will still be logged. Let's file an issue and dig into it, but there's no reason to hold this up over it. In the absence of those errors, we don't want to accidentally cancel outbound snapshots because we've entered into the idle period either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-:-KKLfJVRRJZcyieN8q8m:1417088721)*

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/raft_transport.go, line 268 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKKmkuz4o1XEuYpKS2u-r1-268:-KKPl--MVYeKoF_7jUSz:-1319751378) ([raw file](https://github.com/cockroachdb/cockroach/blob/d95570164a32fc8040e0942f8b307c0456065035/storage/raft_transport.go#L268)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Done.
</details>
This could use a comment. I was about to ask why this only happened on the MsgSnap branch.

---

*[storage/raft_transport.go, line 270 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252#-KKPkvb7pneWjxS2TLB9:-KKPkvb8EzfV5V7ePXCv:849873737) ([raw file](https://github.com/cockroachdb/cockroach/blob/390d63d26cb0897732d6c9b011fc1f2e4c02e546/storage/raft_transport.go#L270)):*
> ```Go
> 			if req.Message.Type == raftpb.MsgSnap {
> 				wg.Add(1)
> 				t.rpcContext.Stopper.RunAsyncTask(func() {
> ```

If RunAsyncTask returns false we need to call wg.Done() to allow the shutdown to complete.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7252)*

**Pull request merged**